### PR TITLE
Version template caching

### LINKED ISSUE #23533: Template path caching causing problems on Atomic sites
Template path caching in core Woocommerce wc_get_template_part() causes Fatal Errors when the WC version changes on Atomic v2.
```
# wp cache get template-part-content-product woocommerce 
/wordpress/plugins/woocommerce/3.6.1/templates/content-product.php 

[25-Apr-2019 17:34:08 UTC] PHP Fatal error: require(): Failed opening required '/wordpress/plugins/woocommerce/3.6.1/templates/content-product.php' (include_path='/:.') in /wordpress/core/5.1.1/wp-includes/template.php on line 706 

xyz.net:/sites/149208628/htdocs# ls -l wp-content/plugins/woocommerce 
lrwxrwxrwx 1 root root 48 Mar 15 07:57 wp-content/plugins/woocommerce -> ../../../../wordpress/plugins/woocommerce/latest 
```
Added in commit d167cb0.

More info on p9o2xV-j7-p2

### PR DESCRIPTION
**Pull request by kloon**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:
https://github.com/woocommerce/woocommerce/commit/d167cb0a9593b8038f510435aa1bd706b7d762cd Introduced template caching to avoid multiple file exist checks on a single page load. However due to symlinks and how you can have multiple versions of a plugin under different folder this caused issues on some hosts which were symlinking WooCommerce based on the version.

This PR adds a version to the template caching which corresponds to the WooCommerce current woocommerce version, if the WooCommerce version changes the cache will be invalid causing it to find the new path.

Another alternative would be to use the plugin file path as the version key and do the validation against that, but templates should not changes in the current version so do not see the need to go that route, just another alternative should we want to.

Closes #23533 

### How to test the changes in this Pull Request:

1. Ensure you have memcache active on your site
2. Load the shop page
3. Rename the WooCommerce plugin folder and change the $version variable in class-woocommerce.php to something else
4. Activate WooCommerce again
5. Load shop page and check error log for any fatal errors, there should be none. 

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Enhancement - Version template caching to avoid fatal errors on symlinked plugins.

### CODE REVIEW

File: includes/wc-core-functions.php [hunk 1]
```diff
@@ -158,7 +158,13 @@ function wc_update_order( $args ) {
  */
 function wc_get_template_part( $slug, $name = '' ) {
 	$cache_key = sanitize_key( implode( '-', array( 'template-part', $slug, $name ) ) );
-	$template  = (string) wp_cache_get( $cache_key, 'woocommerce' );
+	$template  = wp_cache_get( $cache_key, 'woocommerce' );
```
Comment (peterfabian) [inline includes/wc-core-functions.php]: Just wondering if it wouldn't be easier to make the version part of the cache key? Then you wouldn't need to change the rest and if the caching depends on the version anyway, then it perhaps should be part of cache key?

**Review by peterfabian — CHANGES_REQUESTED** [review]

File: includes/wc-core-functions.php (hunk 1 above)
Comment (kloon) [reply includes/wc-core-functions.php]: A good suggestion, will make the changes.

File: includes/wc-core-functions.php [hunk 2]
```diff
@@ -157,7 +157,7 @@ function wc_update_order( $args ) {
  * @param string $name Template name (default: '').
  */
 function wc_get_template_part( $slug, $name = '' ) {
-	$cache_key = sanitize_key( implode( '-', array( 'template-part', $slug, $name ) ) );
+	$cache_key = sanitize_key( implode( '-', array( 'template-part', $slug, $name ) ) . '_' . WC()->version );
```
Comment (peterfabian) [inline includes/wc-core-functions.php]: Thanks for the update. Sorry for nitpicking, but is there a reason to not add the version to the array like the rest of cache_key parts?

**Comment by claudiosanches** [comment]
May I just suggest to change `WC()->version` to `WC_VERSION`?
There 105 results for `WC_VERSION`, and only 21 for `WC()->version`, maybe some time covert it and maybe deprecate it.

**Review by peterfabian — APPROVED** [review]
Works fine. 👍  
Tested with W3 Total Cache object cache and verified the cache for templates behaves as expected (i.e. when the version changed, the cache is refreshed).

**Pull request merged**